Fix multiplayer games with "Qu"

ecrips opened the pull request:
As described in #256 games with "Qu" (or other multi character cells) cannot be successfully shared in multiplayer mode.

The problem is that there is no separator between the cells so the assumption is that there is a single character per cell. This change creates a separator (an underscore) so that we don't have to rely on the assumption.

The drawback is that this will break sharing with older clients (sharing from an older client is supported albeit no attempt is made to handle the "Qu" case so that will still break). Because this introduces an incompatible change I've bumped the versionCode.

A related change will be needed for the web page on https://lexica.github.io/

pserwylo commented:
This is great, thanks so much for the clear and concise description, the two parallel PRs, and the clean and readable code.

When I get a moment in the next few days, I'll be sure to give it a quick test, then bump a new version for F-Droid to build (I release the F-Droid signed binaries to Google Play).

ecrips commented:
The test failure was due to the test attempting to round-trip an old style URL and the new style URL not matching. I've updated the test to avoid round-tripping the old style URL and added a new test for the new style URL

pserwylo commented:
Hi @ecrips. Sorry I haven't gotten around to progressing this. I wanted to stop and think about it properly, which of course means putting aside a block of time. As I was engrossed in some [retrowars](https://github.com/retrowars/retrowars) related work for the last couple of months, I've kept putting this off.

I'm super happy with this contribution, however I was wondering if you could share your thoughts on one alternative and one improvement. After discussing, then we can either merge this as is, add the improvement first, or try something slightly different to fix it. Happy either way, but my preference is usually maintaining backward compatibility of at all possible.

So, proposed improvement:

There is a [`Language.applyMandatorySuffix()`](https://github.com/lexica/lexica/blob/aa0895a6514bb692fbcee4926286e53ac688e926/libraries/trie/src/main/java/com/serwylo/lexica/lang/Language.kt#L30-L36) method which could potentially be used to avoid the need for changing the URL format at all. Do you think this would work? The major caveat of course is that this would not work for the website based version :( 

Proposed enhancement:

I love that you now have a method for bailing when a URL comes from an unsupported version of Lexica. I think this should have a UI built into it so that from the beginning of this feature implementation, users will forever more be notified to upgrade to the latest version (e.g. "Please make sure all players are using the latest version of Lexica" should be enough to cover all scenarios). 

2nd Proposed enhancement:

If we are going to change the URL format, could we base64 encode the letters used in the board? I have found issues playing with my family where there is a tendancy to start thinking about words as soon as you see the URL, because the board is there in plain site. The addition of underscores will make this even more obvious. Base64 encoding the letters would prevent this for the most part (of course people could still decode if they want, but we're not after preventing cheating, just the temptation to start early).

Thoughts? I'd be happy to help out with the two proposed enhancements if you think they are good, or happy for you to do so as well.

ecrips commented:
I've redone the change to Base64 encode the board data and use `applyMandatorySuffix` to accept the older format of the game.
For now I've not redone the changes for the web interface as I see there are changes happening with that in #270 so I'll wait for your feedback before taking a look at that.

In terms of UI: currently there's just a "toast" when the version is incompatible - I've not done any Android UI development in a long time so if you are able to assist then that would be great.

pserwylo commented:
Thanks so much for the quick turnaround. You are correct that great stuff is happening with web lexica right now, but that wont become the official version just yet, so if you were able to apply base64 decoding to the website too, I'll get it merged and then try to release this fix this week.

pserwylo commented:
An I must appologies about the UI request, the toast you have displayed is very informative and will do quite nicely!

pserwylo commented:
Will merge this for now, my offline testing shows it works great (though obviously because I never actually implemented min supported lexica version in the original multiplayer - sending these links to current lexica will break it - nothing we can do about that due to my lack of foresight, but at least it will alert people appropriately for all future versions).

When we get the website working with Base64 (let me know if you don't have the resources to do so and I can look into it), then we'll release that and then bump Lexica to a new version for F-Droid to build.

Thanks for all your help @ecrips.
